src/dataworker/Dataworker.ts

```diff
@@ -1130,10 +1162,72 @@ export class Dataworker {
     );
     if (unexecutedLeaves.length === 0) return;
 
-    // Filter for leaves where the contract has the funding to send the required tokens.
+    // Call `exchangeRateCurrent` on the HubPool before accumulating fees from the executed bundle leaves. This is to
+    // address the situation where `addLiquidity` and `removeLiquidity` have not been called for an L1 token for a
+    // while, which are the other methods that trigger an internal call to `_exchangeRateCurrent()`. Calling
+    // this method triggers a recompounding of fees before new fees come in.
+    await this._updateExchangeRates(unexecutedLeaves, submitExecution);
+
+    // First, execute mainnet pool rebalance leaves. Then try to execute any relayer refund and slow leaves for the
+    // expected relayed root hash, then proceed with remaining pool rebalance leaves. This is an optimization that
+    // takes advantage of the fact that mainnet transfers between HubPool and SpokePool are atomic.
+    const mainnetLeaf = unexecutedLeaves.find((leaf) => leaf.chainId === hubPoolChainId);
+    if (mainnetLeaf) {
+      await this._executePoolRebalanceLeaves(
+        spokePoolClients,
+        [mainnetLeaf],
+        balanceAllocator,
+        expectedTrees.poolRebalanceTree.tree,
+        submitExecution
+      );
```

Comment:
This doesn't happen really in the current system, but just for the sake of correctness, I think there can be multiple pool rebalance leaves for a particular chain.

Thoughts on something more like:
```suggestion
      await this._executePoolRebalanceLeaves(
        spokePoolClients,
        unexecutedLeaves.filter((leaf) => leaf.chainId === hubPoolChainId),
        balanceAllocator,
        expectedTrees.poolRebalanceTree.tree,
        submitExecution
      );
```

Reply:
Yep you're right added this in this commit a11d0b10c768079b4e9cafbcdded9d465cff5ab4